src/engine/enginebufferscalelinear.cpp

```diff
@@ -249,8 +235,14 @@ CSAMPLE* EngineBufferScaleLinear::do_scale(CSAMPLE* buf,
     ceil_sample[0] = 0;
     ceil_sample[1] = 0;
 
+    int samples_read = 0;
     int i = 0;
-    int screwups = 0;
+    //int screwups_debug = 0;
+
+    double rate_add = fabs(rate_old);
```

Comment:
common (for clarity of my question)
```c++
const double rate_diff = rate_new - rate_old;
const double rate_delta = rate_diff / buffer_size;
```

It looks like the old code calculated rate_add as:  

```c++
for (int i = 0; i < buf_size; ++i) {
  const double rate_add = fabs(rate_old + i*rate_delta);
  m_dNextFrame = m_dCurrentFrame + rate_add;
}
``` 
and the new code sums rate add as:

```c++
double rate_add = fabs(rate_old);
const double rate_delta_abs = rate_old > 0 ? rate_delta : -rate_delta;
for (int i = 0; i < buf_size; ++i) {
  m_dNextFrame = m_dCurrentFrame + rate_add;
  rate_add += rate_delta_abs;
}
```

Could you explain what rate_delta_abs is doing? It looks like you're trying to split the ```fabs(rate_old + i*rate_delta)``` into two separate terms to avoid a fabs in the loop. I'm not sure this works in all cases:

Suppose:
```c++
rate_old = -0.5
rate_delta = 1
```
Old way:
```c++
LOOP i = 0
rate_add = 0.5
m_dNextFrame = m_dCurrentFrame + 0.5
LOOP i = 1
rate_add = 0.5 
m_dNextFrame = m_dCurrentFrame + 0.5
```

New way:
```
rate_add = 0.5
rate_delta_abs = -1
LOOP i = 0
m_dNextFrame = m_dCurrentFrame + 0.5
rate_add = 0.5 + -1 = -0.5   // it went negative
LOOP i = 1
m_dNextFrame = m_dCurrentFrame - 0.5 // we interpolated backwards!
```

But maybe I'm missing something. Could you take a look?